src/node.ts

```diff
@@ -29,21 +38,34 @@ export abstract class Node {
   rebindRight(right: Node): Node {
     throw new Error(ERR_NOT_IMPLEMENTED);
   }
+
+  applyHash(root: IHashObject): void {
+    this.h0 = root.h0;
+    this.h1 = root.h1;
+    this.h2 = root.h2;
+    this.h3 = root.h3;
+    this.h4 = root.h4;
+    this.h5 = root.h5;
+    this.h6 = root.h6;
+    this.h7 = root.h7;
+  }
 }
 
 export class BranchNode extends Node {
-  private _root: Uint8Array | null = null;
+  // private _root: Uint8Array | null = null;
+  private hasRoot = false;
 
   constructor(private _left: Node, private _right: Node) {
     super();
     if (!_left || !_right) throw new Error(ERR_INVALID_TREE);
   }
 
-  get root(): Uint8Array {
-    if (!this._root) {
-      this._root = hash(this.left.root, this.right.root);
+  get root(): IHashObject {
```

Comment:
We should maintain compatibility with the existing interface for now. This will make integration w ssz much simpler. (and let us test similarly vs the bigint impl)
See the approach in https://github.com/ChainSafe/persistent-merkle-tree/pull/46 where `root` is just a wrapper for eg: `rootHashObject` that converts from `IHashObject` to `Uint8Array`. All internal calculations can be done using `IHashObject` and we only need to allocate/convert to `Uint8Array` as a final step.